### ISSUE TITLE
Custom dataset is not small enough to be stored in-memory

### ISSUE THREAD

**Issue by camillelhenry2021** [HuguesTHOMAS/KPConv-PyTorch#139]
Hello ! 
Thanks for your open source code.
I'm trying to train KPConv on my own dataset, and I'm having memory problems.
In your S3DIS.py dataloader, I noticed that you were loading all the pointclouds at once, since the dataset is very small. As it is not the case for my dataset, I would like to download my pointclouds one by one, generate the spheres for the downloaded pointcloud, feed the network with it and then when the min_potential of the pointcloud has reached a certain threshold (let's say 2 for example), load another pointcloud and re-iterate the same process. 

Have you already considered this method? Do you think it is feasible? 

Thanks in advance fror your help !

**Comment by HuguesTHOMAS** [comment]
Hi @camillelhenry2021,

I would say it depends on the size of your pointclouds. If you have small point clouds like in the Scannet_v2 dataset or in SemanticKitti, you can even afford to reload the point cloud for every sphere that you pick. In that case, you can adapt the SemanticKitti dataloader.

If your point clouds are very big, then indeed it will be a good idea to actually pick several spheres every time you load one of them. It is totally feasible but you will have to be smart about it because of two things:
1. You still want to have a little bit of every cloud in each epoch (ideally even in each batch) for a better feature learning
2. The dataloader used multiple threads, so you'll need some synchronization when reloading a new point cloud

The first solution (your idea) should be the easiest to implement. You can use the worker_lock that I already have in place to ensure that all the threads are waiting when one of them is reloading a new pointcloud

https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/73e444d486cd6cb56122c3dd410e51c734064cfe/datasets/S3DIS.py#L280

You can then modify the way you choose the `cloud_ind` inside this lock.

https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/73e444d486cd6cb56122c3dd410e51c734064cfe/datasets/S3DIS.py#L296-L298

You can for example create a variable `self.current_cloud_ind` with the index of the currently loaded cloud and another called `self.current_cloud_count` which counts the number of spheres picked in this current cloud. When the count reaches a certain number let's say 100 (but you can tune this depending on your data), you change the current cloud by choosing the new potential minimum: `self.current_cloud_ind = int(torch.argmin(self.min_potentials))` and delete the previous one and load the new one. There might be other parts of the code where you need to change things but you get the idea.

There is a second solution that is better overall but maybe more difficult to implement. You can use a tiling strategy. Divide your point clouds into small square tiles (size just a little bit higher than the sphere diameter). Then every time you pick a sphere, you just have to load 4 tiles (the ones intersecting the spheres). The advantage of this solution is that you will have spheres from different point clouds in each batch (which is not the case with your idea), and also you don't even need to load one full point cloud, which could be useful if your pointclouds are so big that even loading one would cause memory issues. If you are interested in this method, I could give you some advice on what to modify but it would be much more difficult to implement than your idea.

**Comment by camillelhenry2021** [comment]
Hi, 

Thank you for your detailed answer. 

Finally, I decided to do something quite different to avoid memory issues. At each epoch, I load iteratively a sub-part of my training dataset (according to a maximum memory treshold). I train KPConv on this data and I save the checkpoint so that I can reload it while processing the next sub-part. I repeat this process until I have seen all my dataset, then I validate with the same pipeline. I continue this loop until ```config.max_epoch``` parameter is reach.  Theoretically, this process allows me to control how much of the dataset I load in the memory and prevents memory over-consumption. 
Simplifying,  my training process is as follow : 

``` ruby
for epoch in range(config.max_epoch):

   while training_data : # list of my training data
        # load previous checkpoint 

        while memory_storage < memory_limit:
             # load sub-part of my training data

       # process data preparation 
       # train KPConv
       # save current checkpoint 

   # perform validation with same pipeline
````

Surprisingly I still have memory issues with this process. 
        
![Capture](https://user-images.githubusercontent.com/86661421/149177109-080c0c0b-9955-49bf-873c-267211dfabb9.PNG)

Here you can see my free RAM during the training process. The x-axis corresponds to the number of iterations. I set ```config.epoch_steps``` to 500 as it is represented by the blue lines (which means 500 iterations for a sub-part of my training dataset). I set a limit of 30G for each sub-part, which is not much compared to my total RAM capacity which is 126G.  You can see that the memory is approximately released after each sub-part of the dataset, but for some the memory drop suddenly, which is curious because all the sub-part should have roughly the same size. The same behavior occurs for the validation step.

While searching for the cause of this memory leak, I realise that when I stop using data parralelization (ie ```config.input_thread = 1```), I see that my memory is much more stable during the process. You can see the evolution of my free RAM during the validation step below : as expected, the memory jumps and drops roughly the same for each sub-part of my validation dataset. (You can ignore the end of the graph I just stop the process).

![Capture](https://user-images.githubusercontent.com/86661421/149172072-69869d4b-e229-409f-9235-d415812d09cc.PNG)

Because data parralelization saves computing time, I would like to use it in my process, but I have no idea with it makes my RAM unstable. Do you have any idea ?

**Comment by HuguesTHOMAS** [comment]
First, I am not sure if the data is copied for each thread or if the data is shared between threads. That could be something to investigate (see Pytorch dataset class and the behavior when using multiple workers).

However, if that was the case, we would see memory drop at each iteration... In my opinion, this erratic behavior could be caused by the way the python garbage collector works when multiple threads are at play. I never looked into it, but you could try to investigate that. When automatically freeing unused memory, python might be confused by the fact that multiple threads are accessing the data and then maybe wait longer to free the memory. This is just a guess.

**Comment by liuxinren456852** [comment]
> Hi @camillelhenry2021,
> 
> I would say it depends on the size of your pointclouds. If you have small point clouds like in the Scannet_v2 dataset or in SemanticKitti, you can even afford to reload the point cloud for every sphere that you pick. In that case, you can adapt the SemanticKitti dataloader.
> 
> If your point clouds are very big, then indeed it will be a good idea to actually pick several spheres every time you load one of them. It is totally feasible but you will have to be smart about it because of two things:
> 
> 1. You still want to have a little bit of every cloud in each epoch (ideally even in each batch) for a better feature learning
> 2. The dataloader used multiple threads, so you'll need some synchronization when reloading a new point cloud
> 
> The first solution (your idea) should be the easiest to implement. You can use the worker_lock that I already have in place to ensure that all the threads are waiting when one of them is reloading a new pointcloud
> 
> https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/73e444d486cd6cb56122c3dd410e51c734064cfe/datasets/S3DIS.py#L280
> 
> You can then modify the way you choose the `cloud_ind` inside this lock.
> 
> https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/73e444d486cd6cb56122c3dd410e51c734064cfe/datasets/S3DIS.py#L296-L298
> 
> You can for example create a variable `self.current_cloud_ind` with the index of the currently loaded cloud and another called `self.current_cloud_count` which counts the number of spheres picked in this current cloud. When the count reaches a certain number let's say 100 (but you can tune this depending on your data), you change the current cloud by choosing the new potential minimum: `self.current_cloud_ind = int(torch.argmin(self.min_potentials))` and delete the previous one and load the new one. There might be other parts of the code where you need to change things but you get the idea.
> 
> There is a second solution that is better overall but maybe more difficult to implement. You can use a tiling strategy. Divide your point clouds into small square tiles (size just a little bit higher than the sphere diameter). Then every time you pick a sphere, you just have to load 4 tiles (the ones intersecting the spheres). The advantage of this solution is that you will have spheres from different point clouds in each batch (which is not the case with your idea), and also you don't even need to load one full point cloud, which could be useful if your pointclouds are so big that even loading one would cause memory issues. If you are interested in this method, I could give you some advice on what to modify but it would be much more difficult to implement than your idea.

Hi @HuguesTHOMAS  I‘m interested in the second solution，can you give me some advice  on what to modify？Thank you very much.

**Comment by HuguesTHOMAS** [comment]
Hi @liuxinren456852,

Well, first you have to separate your data into square tiles and save each tile as a separate *.ply* file. Use the common formula for grid subsampling: for any point `(x, y, z)`, `X_tile = floor(x / dl)` and `Y_tile = floor(y / dl)`. `X_tile` and `Y_tile` are signed integers that define which tile the point belongs to. `dl` is the size of a tile in meters. Use a naming convention that helps load it easily. For example, something in the likes of: `tile_[dl]_[X_tile]_[Y_tile].ply`. Choose your dl wisely because you do not want to redo this tiling process every time, I would advise something close to half of the radius of the spheres you want to pick.

Then you have to create a function that is able to load specific parts of your data. It would take a center and a radius as input, and load every tile that intersects with the corresponding sphere. Then in the get_item function, you would use this function to load the points you need at run time.

Until now nothing really too difficult. But the hard parts come when you think about the potential sampling. Two cases:
 1. You are able to have a potential array in memory for all the points in your data. In that case, things would work the same and you would not need to modify much of the code
 2. Your data is so big that even having 1 value for every point is too much. In that case, you need to save the potentials in temporary binary files (one for each tile), and load/save them along with the tiles. This means you also need to keep in memory an array of the values of the minimum of potential for each tile (like I already do for multiple Areas in S3DIS)

There should be other things that would need slight modifications in the implementation. Follow the code execution and try to check where you will need to modify. If you have problems, feel free to ask here
